### PR TITLE
Rewrite merge dictionaries

### LINKED ISSUE #321: Rewrite merge dictionaries
In Python 3.9 or greater (see  PEP-584):

```python
x = {"a":1}
y = {"b":2}
z = x | y
```

In Python 3.5 or greater:

```python
x = {"a":1}
y = {"b":2}
z = {**x, **y}
```

In Python 3.4 or greater:

```python
x = {"a":1}
y = {"b":2}
z = dict(**x, **y)
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #321

### CODE REVIEW

File: tests/features/typing_584_test.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+            id='Simple dict rewrite',
+        ),
+        pytest.param(
+            'x = {**a}\n',
+
+            'x = a\n',
+
+            id='Merge of only one dict',
+        ),
```
Comment (asottile) [inline tests/features/typing_584_test.py:42]: this one isn't correct -- the _before_ is a shallow copy and the after is a direct reference

Comment (MarcoGorelli) [reply tests/features/typing_584_test.py]: Thanks, have updated!

Is there any drawback to using `Token('DUMMY', '')` or is `del tokens[i]` preferable?

Comment (asottile) [reply tests/features/typing_584_test.py]: ah so, the reason for those is if it were to process things in reverse, I think you can just `del` those in this case.

an aside: I just updated my repositories to replace "dummy" with "placeholder" (or other equivalent language) to make the code more inclusive

**Review by asottile — APPROVED** [review]

File: tests/features/typing_584_test.py (hunk 1 above)
Comment (MarcoGorelli) [reply tests/features/typing_584_test.py]: I see - I was having issues here when using `del`. Not sure I understand why `del` works at all actually if it changes the number of tokens, but I'll make sense of it

>  I just updated my repositories to replace "dummy" with "placeholder" (or other equivalent language) to make the code more inclusive

Thanks for doing that 👍 

And thanks for your review/merge!

**Pull request merged**